drivers/entropy: stm32 add seed error recovery

str4t0m opened the pull request:
After this changes are applied the driver tries to recover when a seed error is detected.
Series which support the "Conditioning soft reset" perform a soft-reset.
Other series clear the flag and flush the pipeline.

Additionally, in case a zero value is read the value is not used and an error returned instead.

str4t0m committed:
drivers/entropy: stm32 add seed error recovery

This commit introduces an automatic recovery procedure in cases an
entropy source error was detected.
- On Series with soft reset support a soft rest is executed.
- On Series w/o soft reset support the pipeline is cleared by reading
  the RNG_DR 12 times.

With this changes the check for seed errors uses SEIS flag instead
of the SECS flag.

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

ABOSTM commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -114,19 +119,49 @@ static int entropy_stm32_got_error(RNG_TypeDef *rng)
 		} else {
-			retval = LL_RNG_ReadRandData32(
-						    entropy_stm32_rng_data.rng);
+			retval = LL_RNG_ReadRandData32(rng);
```
According to Reference Manual RM0438 (stm32l5):

> Note: When data is not ready (DRDY=”0”) RNG_DR returns zero.
> **It is recommended to always verify that RNG_DR is different from zero**. Because when it is
> the case a seed error occurred between RNG_SR polling and RND_DR output reading (rare
> event).

So it is necessary to check **retval** and if zero,  call recover_seed_error() or return an error.

str4t0m replied:
Thanks. Checking several RMs, this seems to apply to most series, some series don't call it out, but don't list any different possibly invalid reg-value. So I will apply this for all series and push the changes together with the requests above.

ABOSTM reviewed (changes requested):
@str4t0m thanks for this Pull Request.

fabiobaltieri commented:
@ABOSTM would it be possible to rootcause the source of the SECS condition? The manual is not that clear about it, wondering if you had an internal contact that could know more.

Only seen it happening on the WL, maybe if there's an errata about it or some other condition that it impacting the quality of entropy data we could do something to stop it from triggering in the first place.

str4t0m commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -114,19 +119,49 @@ static int entropy_stm32_got_error(RNG_TypeDef *rng)
+	 * This typically takes: 2 AHB clock cycles + 2 RNG clock cycles.
+	 */
+
+	while (LL_RNG_IsEnabledCondReset(rng) ||
```
Thanks very much for this feedback. I checked for SEIS in an earlier draft but removed it because the code never uses SEIS and the interrupt is masked. 

Reading the RM of stm32wl the following is not completely clear to me: Assuming a seed error occurs setting both SEIS and SECS. Would it be possible that the RNG recovers a seed error(SECS reset to 0) without an explicit trigger of CONDRST but there is still data with reduced entropy in the pipeline? Without knowing this for sure it would be probably safer to check for SEIS instead and initiate a soft reset. Can you confirm that if data is ready and SECS is 0, we can be sure the data we read has enough entropy?

Pretty much the same question applies to SOCS without CONDRST support.
The RM suggests to discard 12 words, in order to flush the buffer, as a recovery procedure after initialization.
In case it would be possible that there is data with reduced entropy in the pipeline while SECS is 0, I believe we should add the recovery described in the RM (e.g. RM0444 19.3.7) to recover_seed_error().

It would be great if you have more insight on this, just to be on the safe side. Or is this already ruled out in the RM and I have just overlooked it? I will add the check on SEIS as 2nd one, but I believe the sequence is already respected.

str4t0m replied:
> Would it be possible that the RNG recovers a seed error(SECS reset to 0) without an explicit trigger of CONDRST but there is still data with reduced entropy in the pipeline?
> Yes
Thanks for checking this.
I have now updated the code to check SEIS instead of SECS and either make a soft reset if available or read 12 words otherwise.

ABOSTM commented:
> @ABOSTM would it be possible to rootcause the source of the SECS condition? The manual is not that clear about it, wondering if you had an internal contact that could know more.

I agree with you it looks abnormal and I don't have such issue with RNG example from CubeWL.
Could you open a separate issue so that is can be analyse/review/merged separately.

ABOSTM commented:
> Only seen it happening on the WL, maybe if there's an errata about it or some other condition that it impacting the quality of entropy data we could do something to stop it from triggering in the first place.

It seems the problem is in Zephyr implementation:
probably missing (at least) that part of the driver
https://github.com/zephyrproject-rtos/hal_stm32/blob/164ee6121fb76c9c4f88a087c66c231cee220d8a/stm32cube/stm32wlxx/drivers/src/stm32wlxx_hal_rng.c#L212

str4t0m committed:
drivers: entropy: stm32: don't use zero value

According to the Reference Manual of several series(G0,L5,WL,WB,...)
RNG_DR register value should only be used if it is different from 0:
"Because when it is the case a seed error occurred between RNG_SR
polling and RND_DR output reading (rare event)."

Signed-off-by: Thomas Stranger <thomas.stranger@outlook.com>

str4t0m commented:
Thanks @ABOSTM. After writing 0x0000AA74 to RNG_HTCR these frequent seed errors don't occur anymore.
Do you have the intention to open a PR for this?

str4t0m commented on drivers/entropy/entropy_stm32.c:
```diff
@@ -30,6 +30,11 @@
 #define IRQN		DT_INST_IRQN(0)
 #define IRQ_PRIO	DT_INST_IRQ(0, priority)
 
+#if defined(CONFIG_SOC_SERIES_STM32L5X) || \
```
Since the last push I am using `#if defined(RNG_CR_CONDRST)`, but now I think it would probably be even better to introduce new rng_v3 bindings instead.
Then we could add new properties which include the `RNG_HTCR HTCFG` initialization value and could possibly include a property to disable NIST compliance, set `CLKDIV`, and `RNG_CONFIGX` values.
Should we leave the condition on `RNG_CR_CONDRST` for now and introduce such new bindings in a future PR?

ABOSTM replied:
I am not fully convinced about DT in this case: in all cases , driver need to implement former rng and rev 3.x .
Knowing that there is all we need in 
* CMSIS : RNG_CR_CONDRST which could eventually be replaced by (RNG_VER_3_2 || RNG_VER_3_1 || RNG_VER_3_0) to have explicit reference to REV3
* HAL : RNG_HTCFG  and RNG_HTCFG_1

So such information in DT would be redundant (adding complexity and risk of mistakes).
Other configurations like "disable NIST compliance, set CLKDIV, and RNG_CONFIGX values" are more optional and probably concerns only few experts. So I would leave it for further improvements. 

But let's wait @erwango feedback.

erwango replied:
I do agree with @ABOSTM on this.
This configuration options are quite specific, alternative configurations are mentioned as not tested (so "use at your own risk"), with no benefit documented and not certifiable.
So I don't think we should get into that "complexity" today, without a clearly established use case and expressed need.

str4t0m replied:
I didn't know whether there would be any good reason to choose any different configuration, but I do now agree that this ideally should simply work in a standard compliant configuration.
If you are fine with it, I'd leave it with `#if defined(RNG_CR_CONDRST)`

ABOSTM commented:
> Thanks @ABOSTM. After writing 0x0000AA74 to RNG_HTCR these frequent seed errors don't occur anymore.
> Do you have the intention to open a PR for this?

Yes, I will, but I need to check whether it applies to other series

ABOSTM reviewed (approved):
LGTM
